File: hw/aymacana/u2/hw19Inheritance/FarmSystemJava/src/main/java/ec/edu/espe/farmsystem/model/FarmAnimal.java
package ec.edu.espe.farmsystem.model;

import java.util.Date;

/**
 *
 * @author Mateo Aymacaña, T.A.P. (The Art of Programming), @ESPE
 */
public abstract class FarmAnimal {

    private int id;
    private String breet;
    private Date bornOn;
    private String gender;
    private boolean isAbleToReproduce;
    private float weight;
    private Cage cage;

    public int getAgeInMonths() {
        //TODO compute the age in months
        return 0;
    }

    public void assignCage(Cage cage) {
        this.setCage(cage);
    }

    public FarmAnimal(int id, String breet, Date bornOn, String gender, boolean isAbleToReproduce, float weight, Cage cage) {
        this.id = id;
        this.breet = breet;
        this.bornOn = bornOn;
        this.gender = gender;
        this.isAbleToReproduce = isAbleToReproduce;
        this.weight = weight;
        this.cage = cage;
    }

    @Override
    public String toString() {
        return "FarmAnimal{" + "id=" + id + ", breet=" + breet + ", bornOn=" + bornOn + ", gender=" + gender + ", isAbleToReproduce=" + isAbleToReproduce + ", weight=" + weight + ", cage=" + cage + '}';
    }

    /**
     * @return the id
     */
    public int getId() {
        return id;
    }

    /**
     * @param id the id to set
     */
    public void setId(int id) {
        this.id = id;
    }

    /**
     * @return the breet
     */
    public String getBreet() {
        return breet;
    }

    /**
     * @param breet the breet to set
     */
    public void setBreet(String breet) {
        this.breet = breet;
    }

    /**
     * @return the bornOn
     */
    public Date getBornOn() {
        return bornOn;
    }

    /**
     * @param bornOn the bornOn to set
     */
    public void setBornOn(Date bornOn) {
        this.bornOn = bornOn;
    }

    /**
     * @return the gender
     */
    public String getGender() {
        return gender;
    }

    /**
     * @param gender the gender to set
     */
    public void setGender(String gender) {
        this.gender = gender;
    }

    /**
     * @return the isAbleToReproduce
     */
    public boolean isIsAbleToReproduce() {
        return isAbleToReproduce;
    }

    /**
     * @param isAbleToReproduce the isAbleToReproduce to set
     */
    public void setIsAbleToReproduce(boolean isAbleToReproduce) {
        this.isAbleToReproduce = isAbleToReproduce;
    }

    /**
     * @return the weight
     */
    public float getWeight() {
        return weight;
    }

    /**
     * @param weight the weight to set
     */
    public void setWeight(float weight) {
        this.weight = weight;
    }

    /**
     * @return the cage
     */
    public Cage getCage() {
        return cage;
    }

    /**
     * @param cage the cage to set
     */
    public void setCage(Cage cage) {
        this.cage = cage;
    }

}

File: hw/aymacana/u2/hw19Inheritance/FarmSystemPY/FarmAnimal.py
from abc import ABC, abstractmethod
from datetime import datetime
from Cage import Cage

class FarmAnimal(ABC):
    def __init__(self, id: int, breed: str, born_on: datetime, gender: str, 
                 is_able_to_reproduce: bool, weight: float, cage: Cage):
        self._id = id
        self._breed = breed
        self._born_on = born_on
        self._gender = gender
        self._is_able_to_reproduce = is_able_to_reproduce
        self._weight = weight
        self._cage = cage
    
    def get_age_in_months(self) -> int:
        # TODO compute the age in months
        return 0
    
    def assign_cage(self, cage: Cage) -> None:
        self._cage = cage
    
    # Getters
    @property
    def id(self) -> int:
        return self._id
    
    @property
    def breed(self) -> str:
        return self._breed
    
    @property
    def born_on(self) -> datetime:
        return self._born_on
    
    @property
    def gender(self) -> str:
        return self._gender
    
    @property
    def is_able_to_reproduce(self) -> bool:
        return self._is_able_to_reproduce
    
    @property
    def weight(self) -> float:
        return self._weight
    
    @property
    def cage(self) -> Cage:
        return self._cage
    
    # Setters
    @id.setter
    def id(self, id: int) -> None:
        self._id = id
    
    @breed.setter
    def breed(self, breed: str) -> None:
        self._breed = breed
    
    @born_on.setter
    def born_on(self, born_on: datetime) -> None:
        self._born_on = born_on
    
    @gender.setter
    def gender(self, gender: str) -> None:
        self._gender = gender
    
    @is_able_to_reproduce.setter
    def is_able_to_reproduce(self, is_able_to_reproduce: bool) -> None:
        self._is_able_to_reproduce = is_able_to_reproduce
    
    @weight.setter
    def weight(self, weight: float) -> None:
        self._weight = weight
    
    @cage.setter
    def cage(self, cage: Cage) -> None:
        self._cage = cage
    
    def __str__(self) -> str:
        return (f"FarmAnimal{{id={self._id}, breed='{self._breed}', "
                f"born_on={self._born_on}, gender='{self._gender}', "
                f"is_able_to_reproduce={self._is_able_to_reproduce}, "
                f"weight={self._weight}, cage={self._cage}}}")